### DISCUSSION TITLE
Does get<std::string>() allocate memory?

### DISCUSSION THREAD

**Discussion by Jack012a** [nlohmann/json#3835] [Q&A] [answered]
Hello,

Which of the following method to retrieve the std::string value is better?

```
auto v1 = json_types.at["/field_name/"_json_pointer].get<std::string>();
const auto& v2 = json_types.at["/field_name/"_json_pointer].get<std::string>();
```

My understanding is that if get<std::string>() allocates memory, then both methods have similar performance.
However, if get<std::string>() doesn't allocate memory, then method two is better because it doesn't involve an extra copy.

Thank you

**Comment by falbrechtskirchinger** [comment, marked as answer]
Use `get_ref<const std::string&>()` to avoid allocations.
Note that JSON pointers aren't cheap either.

**Reply by Jack012a** [reply]
@falbrechtskirchinger 

I use this method a lot in my application:

`auto v1 = json_types.at["/top_field/second_field"_json_pointer].get<std::string>();`

So your suggestion is that I should use the following instead?
`auto v1 = json_types.at["top_field"]["second_field"].get_ref<const std::string&>();`

I want to use the most efficient way to retrieve the value.

Thank you

**Reply by falbrechtskirchinger** [reply]
Almost.
```cpp
const auto &v1 = json_types["top_field"]["second_field"].get_ref<const std::string&>();
```

Assigning to `auto v1` would create a copy. (Also, `at[]` isn't a thing. There're `operator[]` and `at()`.)

**Comment by sjanel** [comment]
If your compiler supports C++17, you can use as well (which is simpler) `get<std::string_view>()`
